[Discover] Implement Discover customization framework

davismcphee opened the pull request:
## Summary

This PR includes the initial implementation of the Discover customization framework based on the `2023-04 Discover Customizations` RFC.

![customizations](https://github.com/elastic/kibana/assets/25592674/db8202d1-666d-44da-84b0-dce4fd1118e5)

Notes:
- I've included two initial extension points in this PR: `top_nav` and `search_bar`. To my knowledge, these are the ones o11y want to start with, but we don't yet have product alignment on these decisions. ~~I've left them in for now for testing purposes, but I'll need to update this PR before merging to either add tests for these extensions points if we decide to include them, or remove them if we decide not to include them.~~ Tests have now been added for these customizations.
- I'm planning to open a separate PR with documentation about the framework once this is merged, but merging this first will unblock o11y.
- In order to enable customization profiles, Discover has been updated to user locators for all of its navigation, which will allow the current profile to be maintained when navigating between routes. This is because the current customization profile is stored in the URL path as `/p/{profile_name}/{discover_route}`.

Resolves #158625.

### Checklist

- [ ] ~Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)~
- [ ] ~[Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials~
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] ~Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))~
- [ ] ~If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)~
- [ ] ~This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))~
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

jughosta reviewed (commented):
Great job, Davis!

kertal commented on src/plugins/discover/public/application/main/components/top_nav/get_top_nav_links.tsx:
```diff
@@ -85,7 +87,7 @@ export const getTopNavLinks = ({
       defaultMessage: 'New Search',
     }),
-    run: () => navigateTo('/'),
+    run: () => services.locator.navigate({}),
```
I like this change ... just saying ...

kertal commented on src/plugins/discover/public/application/main/components/top_nav/get_top_nav_links.tsx:
```diff
@@ -224,17 +237,41 @@ export const getTopNavLinks = ({
-    shareSearch,
-    ...(services.triggersActionsUi &&
+  const defaultMenu = topNavCustomization?.defaultMenu;
+  const entries = topNavCustomization?.getMenuItems?.() ?? [];
```
I'd assume it's fine here because the function is returning a new array and not a ref of an existing one?

kertal commented on src/plugins/discover/public/application/main/components/top_nav/get_top_nav_links.tsx:
```diff
@@ -224,17 +237,41 @@ export const getTopNavLinks = ({
+  }
+
+  if (!defaultMenu?.new?.disabled) {
+    entries.push({ data: newSearch, order: defaultMenu?.new?.order ?? 200 });
+  }
+
+  if (!defaultMenu?.open?.disabled) {
+    entries.push({ data: openSearch, order: defaultMenu?.open?.order ?? 300 });
+  }
+
+  if (!defaultMenu?.share?.disabled) {
+    entries.push({ data: shareSearch, order: defaultMenu?.share?.order ?? 400 });
+  }
+
+  if (
+    services.triggersActionsUi &&
     services.capabilities.management?.insightsAndAlerting?.triggersActions &&
-    !isPlainRecord
-      ? [alerts]
-      : []),
-    inspectSearch,
-    ...(services.capabilities.discover.save ? [saveSearch] : []),
-  ];
+    !isPlainRecord &&
+    !defaultMenu?.alerts?.disabled
+  ) {
+    entries.push({ data: alerts, order: defaultMenu?.alerts?.order ?? 500 });
+  }
+
+  if (!defaultMenu?.inspect?.disabled) {
+    entries.push({ data: inspectSearch, order: defaultMenu?.inspect?.order ?? 600 });
+  }
+
+  if (services.capabilities.discover.save && !defaultMenu?.save?.disabled) {
+    entries.push({ data: saveSearch, order: defaultMenu?.save?.order ?? 700 });
+  }
```
apart from the added functionality, I like that it's more readable now 👍

kertal reviewed (commented):
Finished my first round of review. looking pretty good 👏  nice changes under the hood, working as expected, left a few minor comments, and will have another look, but now I'm hungry 🍣

davismcphee commented on src/plugins/discover/public/customizations/customization_service.ts:
```diff
@@ -0,0 +1,73 @@
+  const customizations = new Map<DiscoverCustomizationId, CustomizationEntry>();
+
+  return {
+    set: (customization) => {
```
The arguments have implicit typing from the interface, but explicit typing never hurts, updated here: https://github.com/elastic/kibana/pull/158603/commits/7c76433244db34f0eba050ba555b8886255be8d5

davismcphee commented on src/plugins/discover/public/customizations/customization_types/top_nav_customization.ts:
```diff
@@ -0,0 +1,35 @@
+  share?: TopNavDefaultMenuItem;
+  alerts?: TopNavDefaultMenuItem;
+  inspect?: TopNavDefaultMenuItem;
+  save?: TopNavDefaultMenuItem;
```
That's fair, added `Item` suffixes here: https://github.com/elastic/kibana/pull/158603/commits/0d4e97b4a8e34e947e57173d29f6749fff734788

davismcphee commented on src/plugins/discover/public/application/main/discover_main_app.tsx:
```diff
@@ -70,8 +63,8 @@ export function DiscoverMainApp(props: DiscoverMainProps) {
-    setBreadcrumbsTitle(savedSearch.title, chrome);
-  }, [savedSearch.id, savedSearch.title, chrome, data]);
+    setBreadcrumbsTitle({ title: savedSearch.title, services });
+  }, [savedSearch.id, savedSearch.title, chrome, data, services]);
```
Yup we can! Updated here: https://github.com/elastic/kibana/pull/158603/commits/9c94c50d9da62288be6911ddfcffa76701c86f79

davismcphee commented on examples/discover_customizer/kibana.jsonc:
```diff
@@ -0,0 +1,11 @@
+{
+  "type": "plugin",
+  "id": "@kbn/discover-customizer-plugin",
```
Yeah that's a good point, I updated the plugin to be `discover_customization_examples`: https://github.com/elastic/kibana/pull/158603/commits/e4ccadb0c045a4750370b2c23aaceea4d6792634

davismcphee reviewed (commented):
@jughosta @kertal Thanks for all the feedback! I think I addressed all of the concerns, and I also decided to add some functional and Jest tests for the example plugin and customization points here: https://github.com/elastic/kibana/pull/158603/commits/0daf70389ecc28bac9564422f19a66e2e816f71a. So if we do decide to go with these extension points initially, the tests are now there.

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

kertal commented:
It's clear that the page load bundle needs to increase with the PR, given all those changes that can't be a applied in the async bundle, however, it might be worth to check if it can be optimized to be less than 4 kb ... anyway no blocker
<img width="341" alt="Bildschirmfoto 2023-06-05 um 11 19 02" src="https://github.com/elastic/kibana/assets/463851/0c9df6c6-3fdb-4cfd-ab30-26eb89f4fbc4">

davismcphee commented:
> It's clear that the page load bundle needs to increase with the PR, given all those changes that can't be a applied in the async bundle, however, it might be worth to check if it can be optimized to be less than 4 kb ... anyway no blocker <img alt="Bildschirmfoto 2023-06-05 um 11 19 02" width="341" src="https://user-images.githubusercontent.com/463851/243310652-0c9df6c6-3fdb-4cfd-ab30-26eb89f4fbc4.png">

I looked into this a bit and it seems that ~3.25KB of the 3.9KB comes from importing `createProfileRegistry`, `createCustomizeFunction`, and `ProfileAwareLocator` in `plugin.tsx`. I spent some time trying to optimize things and according to `scripts/build_kibana_platform_plugins`, it looks like I was able to reduce the page load bundle by ~2.5KB by async importing `ProfileAwareLocator` when it's used (https://github.com/elastic/kibana/pull/158603/commits/e347d3eb64b3a00b4927a9463ecd0ad1144d9b38), and changing the `createProfileRegistry` and `createCustomizeFunction` imports a bit (https://github.com/elastic/kibana/pull/158603/commits/ba9eb72bda57b47fe5284c0cad2a50d5be45141a). But the build just started so we'll see what CI says in the morning!

>BTW: not sure we should label it as a feature, because it becomes a feature when solutions are using it

The way I was thinking about it was that the framework is a new feature exposed by Discover, but it's not user facing, so maybe not something we need in the release notes. I changed it to `release_note:skip` now.

kertal commented on src/plugins/discover/public/application/main/components/top_nav/discover_topnav.tsx:
```diff
@@ -108,18 +107,27 @@ export const DiscoverTopNav = ({
     });
   }, [dataViewEditor, stateContainer]);
 
+  const topNavCustomization = useDiscoverCustomization('top_nav');
```
sounds good

kertal commented on src/plugins/discover/public/application/discover_router.tsx:
```diff
@@ -18,36 +18,108 @@ import { DiscoverMainRoute } from './main';
+      <Route path={prefixPath('doc/:dataViewId/:index')}>
+        <SingleDocRoute />
+      </Route>
+      <Route path={prefixPath('viewAlert/:id')}>
+        <ViewAlertRoute />
+      </Route>
```
Don't think this needs to be part of this PR 👍

kertal reviewed (approved):
LGTM; did a quick final code review + test, works as expected 👍  Let's ship it 🚢 
Maybe worth to mention in the description the change in the URL for profile URLs. Because it's a significant change, and worth to be noted ... however you intended to add documentation in a follow up PR, so it's should be covered then.

stratoula commented on src/plugins/unified_search/public/query_string_input/query_bar_top_row.tsx:
```diff
@@ -527,6 +528,7 @@ export const QueryBarTopRow = React.memo(
     }
 
     function renderDataViewsPicker() {
+      if (props.customDataViewPicker) return props.customDataViewPicker;
```
I would prefer this check to not happen here but happen here instead https://github.com/elastic/kibana/blob/main/src/plugins/unified_search/public/query_string_input/query_bar_top_row.tsx#L677


In my mind this is a custom element which is rendered in the position of the dataview picker. In your example I see a saved searches picker (not dataviews) and can be anything else. So I think:
1. Should not be on the renderDataViewPicker function but be a different function instead
2. Change the name, customDataViewPicker indicates to me that I have a dataview picker which just uses a different layout. But this is something else.

davismcphee replied:
Makes sense to me -- I renamed the prop to `dataViewPickerOverride` instead and moved the check to where you suggested: https://github.com/elastic/kibana/pull/158603/commits/fc8d15e0c6a0c41b0bf62d1d99740be8280a625f. Let me know if that seems good!

jughosta commented:
Hi Davis,
There is one more case for Data View Picker customization in Discover: in mobile view we show fields in a flyout like this:
https://github.com/davismcphee/kibana/blob/discover-customizations/src/plugins/discover/public/application/main/components/sidebar/discover_sidebar.tsx#L280

<img width="500" alt="Screenshot 2023-06-13 at 18 13 07" src="https://github.com/elastic/kibana/assets/1415710/42b95f82-8844-4abb-9eb9-071df5475e63">

davismcphee commented:
> LGTM; did a quick final code review + test, works as expected 👍 Let's ship it 🚢 Maybe worth to mention in the description the change in the URL for profile URLs. Because it's a significant change, and worth to be noted ... however you intended to add documentation in a follow up PR, so it's should be covered then.

I added a note to the description about URL and locator changes to support profiles, and I'll expand on it more in the followup documentation PR.

> Hi Davis, There is one more case for Data View Picker customization in Discover: in mobile view we show fields in a flyout like this and it's handled separately in the code: https://github.com/davismcphee/kibana/blob/discover-customizations/src/plugins/discover/public/application/main/components/sidebar/discover_sidebar.tsx#L280
> 
> <img alt="Screenshot 2023-06-13 at 18 13 07" width="500" src="https://user-images.githubusercontent.com/1415710/245541085-42b95f82-8844-4abb-9eb9-071df5475e63.png">

Great catch, thanks! I added support and tests here: https://github.com/elastic/kibana/pull/158603/commits/e6b9cf9d8823b4145ff3c2d19f4b2e45df764d58.

stratoula reviewed (approved):
Thanx Davis, this looks fantastic and I like the new prop name too! LGTM!

jughosta commented:
Thanks, Davis 👍
